---
title: "DESeq analysis"
author: "Natalia Andrade and Ira Cooke"
date: "07/08/2017"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(cache=TRUE)
options(width = 60)

# Run if needed to install ComplexHeatmap
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("ComplexHeatmap")

library(knitr)
library(DESeq2)
library(tidyverse)
library(ggrepel)
library(ComplexHeatmap)
library(ggpubr)
library(colorspace)
```

Counts obtained from Corset were analysed with DESeq to identify differentially expressed genes between HardCoral treatments. 

```{r, include=FALSE}
#First we prepare the data.  
#This involves reading in a counts matrix from corset.  
#Then we extract the sample names from the columns and match those to conditions in an excel file that describes the experimental conditions for each sample.

counts <- read_tsv("hpc/corset/03-counts.txt") %>% 
  column_to_rownames(var="X1") %>% 
  as.matrix()

sample_data <- read_csv("raw_data/Samples_data.csv",trim_ws = TRUE)
colnames(sample_data) <- c("ID","Tank","HardCoral",'treat',"SoftCoralControl","PdvsL_Other")

sample_data_row_order <- match(colnames(counts),sample_data$ID)
```

Initial data exploration with PCA revealed that PdLd is an extreme outlier.  We therefore excluded this sample from further analysis

```{r, include=FALSE}
dds <- DESeqDataSetFromMatrix(countData = counts,colData=sample_data[sample_data_row_order,],design = ~  treat+ HardCoral)
vsd <- varianceStabilizingTransformation(dds, blind=FALSE)

pca_data <- plotPCA(vsd,intgroup=c("HardCoral","SoftCoralControl"),returnData=TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"))

pca_data %>% 
  ggplot(aes(x=PC1,y=PC2)) + 
  geom_point(aes(color=SoftCoralControl,shape=HardCoral),size=4) + 
  geom_label_repel(aes(label=group)) + 
  xlab(paste0("PC1: ",percentVar[1],"% variance")) + 
  ylab(paste0("PC2: ",percentVar[2],"% variance"))  +
  theme(text = element_text(size=20))
```

Differential Expression Analysis with the outlier excluded was then performed based on a model with no intercept and a separate coefficient for each level of the variable `HardCoralTrt`.  This model gives us maximum flexibility for statistical testing using different contrasts.

> ~ 0+HardCoralTrt

```{r}
counts2 <- data.frame(counts) %>% select(-PdLd) %>% as.matrix()

sample_data2 <- sample_data[match(colnames(counts2),sample_data$ID),] %>% 
  unite("HardCoralTrt",HardCoral,treat,remove = FALSE)

dds2 <- DESeqDataSetFromMatrix(countData = counts2,colData=sample_data2,design = ~  0+HardCoralTrt)

if (file.exists("cache/dds2.rds")) {
    dds2 <- read_rds("cache/dds2.rds")
} else {
  dds2 <- DESeq(dds2, parallel = TRUE)
  write_rds(dds2,"cache/dds2.rds")
}

resultsNames(dds2)
```

Before proceeding with contrasts we first check a PCA with the outlier excluded. This shows very clearly that Porites genotype is the dominant source of variation in the data, accounting for 89%. PC2 accounts for just 5% of variation and captures differences due to the competing Lobophytum colony and whether competition exists at all (Control).

```{r}
vsd2 <- varianceStabilizingTransformation(dds2, blind=FALSE)
pca_data2 <- plotPCA(vsd2,intgroup=c("HardCoral","SoftCoralControl"),returnData=TRUE)
percentVar2 <- round(100 * attr(pca_data2, "percentVar"))

pca_data2 %>% 
  ggplot(aes(x=PC1,y=PC2)) + 
  geom_point(aes(color=SoftCoralControl,shape=HardCoral),size=4) + 
  geom_label_repel(aes(label=group))  + 
  xlab(paste0("PC1: ",percentVar2[1],"% variance")) + 
  ylab(paste0("PC2: ",percentVar2[2],"% variance"))  +  
  theme(text = element_text(size=20)) + 
  theme_pubclean() +
  theme(legend.position = "right")
ggsave("figures/pca.png", width = 12, height = 8)
```

To set up contrasts we first use `resultsNames()` to extract a list of fitted coefficients in the model

```{r}
resultsNames(dds2)
```

Based on this we formulate the contrasts;

1. `contrast_ct <- c(1,-1,1,-1)` which captures genes consistently different between treatment and control across both Porites colonies
2. `contrast_pd <- c(1,-1,0,0)` capturing genes DE between treatment and control for `Pd` only
3. `contrast_pf <- c(0,0,1,-1)` capturing genes DE between treatment and control for `Pf` only


```{r}
#HardCoralTrtPd_C" "HardCoralTrtPd_T" "HardCoralTrtPf_C" "HardCoralTrtPf_T"

# This contrast gives genes consistently DE between treatment and control
 res_ct <- results(dds2,contrast = list(c("HardCoralTrtPd_C","HardCoralTrtPf_C"),c("HardCoralTrtPd_T","HardCoralTrtPf_T")),listValues=c(1/2,
 -1/2), tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1)

contrast_pd <- c(1,-1,0,0)
contrast_pf <- c(0,0,1,-1)

res_pd <- results(dds2,contrast = contrast_pd,tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1)

res_pf <- results(dds2,contrast = contrast_pf,tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1)
```


For the top genes differentially expressed between control and treatment we plot raw values as scatterplots.  This is just to give a feel for how strong these candidates are when looking at the raw data directly.  

Scanning through these plots I can see quite a few genes with consistent differences between both controls and the treatment values ... BUT .. there are others where the change seems to be driven by a strong change in just one of the controls (usually Pf).  This also seems to emerge from looking at the heatmap (see below).

```{r}
# For genes DE ONLY in  contrast btw Control and treatment

top_genes <- res_ct %>% arrange(padj) %>% top_n(24) %>% pull(row)

ct_genes_tidy <- assay(vsd2)[which(rownames(vsd2) %in% res_ct$row),] %>%
  as.data.frame() %>%
  rownames_to_column("cluster_id")

ct_genes_tidy %>%
  filter(cluster_id %in% top_genes) %>% 
  pivot_longer(-cluster_id,names_to = "ID",values_to="count") %>%
  left_join(sample_data) %>%
  ggplot(aes(x=SoftCoralControl, y=log(count), colour=HardCoral)) +
  geom_point(aes(shape=treat)) +
  facet_wrap(~cluster_id,ncol = 3,scales = "free_y")
```

Now a heatmap for all the DE genes between treatment and control. For this the relative change in expression is plotted (relative to the mean for a gene) so that clustering is meaningful. The clusters reveal some interesting patterns (not sure if worth writing about but interesting).  There is a big cluster where the Pf control has higher expression.  There are also clusters driven by higher expression in a few specific interactions (PfLa, PfLc, PdLb) for one cluster .. and (PdLa, PdLe, PfLc) for the other.  It would be fascinating if this was also evident in the behavioural data.  Other clusters are driven by an overall higher expression in Pd vs Pf (bottom small cluster).


```{r}
ct_heatmap_data <- ct_genes_tidy %>%
  as.data.frame() %>%
  column_to_rownames("cluster_id")

ct_heatmap_data_relative <- sweep(ct_heatmap_data, MARGIN=1, STATS= rowMeans(ct_heatmap_data))

colgroup_colours <- qualitative_hcl(3, palette = "Dark 3")

Heatmap(ct_heatmap_data_relative, column_km = 3, row_km = 6, show_row_names = FALSE,column_title_gp = gpar(fill = colgroup_colours, font = 1:3))
```


Finally let's try to create a composite figure including the PCA and heatmap because there are related patterns seen in both figures that we would like to make clear.

```{r}
library(cowplot)
library(grid)

hm_col_groups <- list(down = c("PdLb","PfLd","PfLa"), neutral = c("PdC","PdLc","PfLe","PfLb","PfC"), up = c("PdLe","PdLa","PfLc"))

pca_data3 <- pca_data2 %>% 
  mutate(colgroup = "neutral") %>% 
  mutate(colgroup = ifelse(name %in% hm_col_groups$up,"up","neutral")) %>% 
  mutate(colgroup = ifelse(name %in% hm_col_groups$down,"down",colgroup))

colgroup_colours <- diverging_hcl(3, palette = "Blue-Red 2")

pca_plot <- pca_data3 %>% 
  ggplot(aes(y=PC1,x=PC2)) + 
  geom_point(aes(shape=HardCoral, col=colgroup),size=6) + 
#  geom_point(aes(x=-1,y=PC2, color=colgroup),size=4)+
  geom_label_repel(aes(label=name),point.padding = unit(0.2,"cm"))  + 
  ylab(paste0("PC1: ",percentVar2[1],"% variance")) + 
  xlab(paste0("PC2: ",percentVar2[2],"% variance"))  +  
  scale_color_discrete_diverging(palette = "Blue-Red 2") +
#  scale_color_discrete_qualitative(palette = "Dark 3") +
  theme(text = element_text(size=20)) + 
  theme_pubclean() +
  guides(fill = FALSE,col=FALSE) +
  theme(legend.position = "right", legend.title = element_blank())

library(circlize)
col_fun = colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))

# Perform kmeans clustering and save the result in cache.
# There is some randomness in this so we save it for consistency
# This also allows transcripts within cluster to be extracted in later scripts
#
if ( file.exists("cache/row_km.rds")){
  row_kmeans <- read_rds("cache/row_km.rds")
} else {
  row_kmeans <- kmeans(ct_heatmap_data_relative,centers = 5,nstart = 100)
  write_rds(row_kmeans,"cache/row_km.rds")
}

hm <- Heatmap(ct_heatmap_data_relative, 
                                      col = col_fun,
                                      column_km_repeats = 100,
                                      # row_km_repeats = 100,
                                      column_km = 3,
                                      row_split = row_kmeans$cluster,
                                      show_row_dend = FALSE,
                                      show_row_names = FALSE,
                                      column_title_gp = gpar(fill = colgroup_colours, font = 1:3),
                                      heatmap_legend_param = list(title="LogFC", legend_width = unit(10,"cm")))

hm_plot <- grid.grabExpr(draw(hm))

plot_grid(pca_plot,hm_plot,ncol = 1, 
          rel_heights = c(0.4,0.6), 
          labels = c("A","B"),
          label_x = 0, label_y = 0,
          hjust = -0.5, vjust = -0.5)

ggsave2("figures/pca_hm.png", width = 12,height = 7)
```


```{r}



```


