---
title: "Gene function analysis"
author: "Natalia Andrade and Ira Cooke"
date: "07/08/2017"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(cache=TRUE)
options(width = 60)

library(knitr)
library(DESeq2)
library(tidyverse)
library(ggrepel)
library(pheatmap)
library(UpSetR)
library(ggplot2)
```



Load differentially expressed genes between HardCoral treatments obtain data from analysis with DESeq2 (03_deseq). 

```{r, include=FALSE}
#First we prepare the data.  
#This involves reading in dds2 file from DESeq2 analysis.  
dds2 <- DESeqDataSetFromMatrix(countData = counts2,colData=sample_data2,design = ~  0+HardCoralTrt)

if (file.exists("cache/dds2.rds")) {
    dds2 <- read_rds("cache/dds2.rds")
} else {
  dds2 <- DESeq(dds2)
  write_rds(dds2,"cache/dds2.rds")
} ### if we make this work I think we could just use dds2 <- read_rds("cache/dds2.rds") ###

# We transform the data
vsd2 <- varianceStabilizingTransformation(dds2, blind=FALSE)

contrast_ct <- list(c("HardCoralTrtPd_C","HardCoralTrtPf_C"),c("HardCoralTrtPd_T","HardCoralTrtPf_T"))

#Then we extract the results for the contrast between Control and Treatment
res_ct <- results(dds2,contrast = contrast_ct,listValues=c(1/2,
 -1/2), tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1) # Total 174 genes

#Transformed expression per sample
ct_genes_tidy <- assay(vsd2)[which(rownames(vsd2) %in% res_ct$row),] %>%
  as.data.frame() %>%
  rownames_to_column("cluster_id")

#DESeq2 results and transformed exprestion per sample
DEG_174<- left_join(res_ct, ct_genes_tidy, by=c("row"="cluster_id")) %>% rename("cluster_id"="row")
```

Annotating the 174 DEG_CvsT with transcriptome annotation and manual annotation
```{r annotating DEG_CvsT}

# Transcriptome annotation
P.cylindrica_144087anno<- readRDS("~/Dropbox/natalia_competition/Porites_competition/Annotation/data/P.cylindrica_144087anno.rda") %>% select(-PdLa,-PdLb, -PdLc,-PdLd,-PdLe, -PdC, -PfLa, -PfLb, -PfLc, -PfLd, -PfLe, -PfC)
DEG_174<- inner_join(P.cylindrica_144087anno,DEG_174, by=c("cluster_id"))

# Manual annotation
Manual<- read.csv('/Users/jc275567/Dropbox/Mio/Competition/PORITES/Gene_Function_Analysis/DE_193_database.csv')

DEG_174<- right_join(Manual,DEG_174, by='cluster_id')
#write.csv(DEG_174, '/Users/jc275567/Dropbox/Mio/Competition/PORITES/Gene_Function_Analysis/DE_174_database.csv')
```

Plots showing the expression of genes in the different categories anad for the two genotypes

```{r}
# Graph use in 
boxplot(log10(assays(dds2)[["cooks"]]), range=0, las=2)


# For genes DE ONLY in  contrast btw Control and treatment

top_genes <- DEG_174 %>% arrange(padj) %>% top_n(20) %>% pull(cluster_id)

DEG_174 %>%
  filter(cluster_id %in% top_genes) %>% select(cluster_id,log2FoldChange, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) %>% 
  pivot_longer(c(-cluster_id, -log2FoldChange),names_to = "ID",values_to="count") %>%
  left_join(sample_data) %>% 
  ggplot(aes(x=SoftCoralControl, y=log(count), colour=log2FoldChange)) +
  geom_point(aes(shape=treat)) +
  facet_wrap(~cluster_id,ncol = 3,scales = "free_y")
```

Now a heatmap for all the DE genes between treatment and control. For this the relative change in expression is plotted (relative to the mean for a gene) so that clustering is meaningful. The clusters reveal some interesting patterns (not sure if worth writing about but interesting).  There is a big cluster where the Pf control has higher expression.  There are also clusters driven by higher expression in a few specific interactions (PfLa, PfLc, PdLb) for one cluster .. and (PdLa, PdLe, PfLc) for the other.  It would be fascinating if this was also evident in the behavioural data.  Other clusters are driven by an overall higher expression in Pd vs Pf (bottom small cluster).


```{r}
ct_heatmap_data <- ct_genes_tidy %>%
  as.data.frame() %>%
  column_to_rownames("cluster_id")

ct_heatmap_data_relative <- sweep(ct_heatmap_data, MARGIN=1, STATS= rowMeans(ct_heatmap_data))


pheatmap(ct_heatmap_data_relative, cluster_rows=FALSE, show_rownames=TRUE,cluster_cols=TRUE)
```
