---
title: "Gene function analysis"
author: "Natalia Andrade and Ira Cooke"
date: "07/08/2017"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(cache=TRUE)
options(width = 60)

library(knitr)
library(DESeq2)
library(tidyverse)
library(readxl)
library(ggrepel)
library(ggplot2)
library(cowplot)
# if(!require(devtools)) install.packages("devtools")
# devtools::install_github("krassowski/complex-upset")

#library(ComplexUpset)
```

Gene function analysis is based on the following datasets;
  - Functional annotations created in `01_annotate` for all clusters
  - Differential expression analysis (from `02_deseq.Rmd`) to select genes DE between control and treatment
  - Manual annotations created by curating automatic annotations along with literature searches for DE genes
  - K-means clustering groups which identify genes (Corset clusters) identified in the heatmap (see `04_polyp_activity.Rmd`)
  
```{r, include=FALSE}
# For the purpose of this analysis we first join 3 datasets together. 
# Dataset 1 is the differential expression statistics and normalized counts for all transcripts
# Dataset 2 is the annotation information for P.cylindrica transcripts
# Dataset 3 is a table of manual functional annotations for key genes

#
# Dataset 1
#
if (file.exists("cache/dds2.rds")) {
    dds2 <- read_rds("cache/dds2.rds")
} else {
  stop("Nothing in cache/dds2.rds. You must run 02_DESeq.Rmd first")
}

vsd2 <- varianceStabilizingTransformation(dds2, blind=FALSE)

contrast_ct <- list(c("HardCoralTrtPd_C","HardCoralTrtPf_C"),c("HardCoralTrtPd_T","HardCoralTrtPf_T"))

#Then we extract the results for the contrast between Control and Treatment
res_ct <- results(dds2,contrast = contrast_ct,listValues=c(1/2,
 -1/2), tidy = TRUE) %>% 
  arrange(padj) %>% 
  filter(padj<0.1) # Total 174 genes

# (Old version) Transformed expression per sample (Dataset 1)
# This gives results that are hard to reconcile against logFC when there are many zeros
#
# ct_res_vsd2 <- assay(vsd2)[which(rownames(vsd2) %in% res_ct$row),] %>%
#   as.data.frame() %>%
#   rownames_to_column("cluster_id") %>% 
#   right_join(res_ct, by=c("cluster_id"="row"))


# (New) Normalized counts per sample (Dataset 1)
ct_res_ntd2 <- counts(dds2, normalized=TRUE)[res_ct$row,] %>%
  as.data.frame() %>%
  rownames_to_column("cluster_id") %>% 
  right_join(res_ct, by=c("cluster_id"="row"))
```

```{r}
#
# Dataset 2
#
# Transcriptome annotation 
automatic_annotations <- read_rds("cache/annotated_clusters_Porites.rda")

colnames_to_exclude <- colnames(automatic_annotations)[-1]

# Dataset 3
#
# Manual annotation
manual_annotations <- read_excel("raw_data/DE_174_database.xlsx") %>% 
  select(!any_of(colnames_to_exclude)) %>% 
  select(-log2FoldChange)
```


```{r join_datasets}
# Joining all three datasets for the DEG
#
annotated_DEG_174 <- ct_res_ntd2 %>% 
  left_join(automatic_annotations) %>% 
  left_join(manual_annotations)
```

Our focus initially is on the 174 genes differentially expressed between control and treatment. Raw (normalised) counts for a handful of the top genes are plotted here as a sanity check to ensure that they look genuinely differentially expressed.

```{r}
sample_data <- read_csv("raw_data/Samples_data.csv",trim_ws = TRUE)
colnames(sample_data) <- c("ID","Tank","HardCoral",'treat',"SoftCoralControl","PdvsL_Other")


# For genes DE ONLY in  contrast btw Control and treatment

top_genes <- annotated_DEG_174 %>% slice_min(padj, n=9, with_ties=F) %>%  pull(cluster_id)

annotated_DEG_174 %>%
  filter(cluster_id %in% top_genes) %>% 
  select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) %>% 
  pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
  left_join(sample_data) %>% 
  ggplot(aes(x=treat, y=log(count), colour=HardCoral)) +
  geom_jitter(aes(shape=SoftCoralControl), height=0, width=0.1) +
  facet_wrap(~cluster_id,ncol = 3,scales = "free_y") +
  ggtitle("Top differentially expressed genes in treatment (competition) versus control")
```


## Function of genes in PCA cluster

```{r}
vsd2 <- read_rds("cache/vsd2.rds")
res_ct <- read_rds("cache/res_ct.rds")
row_kmeans <- read_rds("cache/row_km.rds")

write_cluster <- function(clnum){
  cluster_ids <- names(which(row_kmeans$cluster==clnum))
  DE_cl <- annotated_DEG_174 %>% filter(cluster_id %in% cluster_ids)
  write_tsv(DE_cl,paste("cache/DE_cl",clnum,".tsv",sep = "",collapse = ""))  
}

for(i in 1:5){
  write_cluster(i)
}

DE_cl2 <- read_tsv("cache/DE_cl2.tsv")
DE_cl4 <- read_tsv("cache/DE_cl4.tsv")

```



```{r}
# For gene classify as RESISTING related genes

topResistingPorites<- DEG_174 %>% arrange(padj) %>% filter(RESISTING==TRUE) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) 

plot_portiesR<- topResistingPorites %>%  
    pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
    left_join(sample_data) %>% 
    ggplot(aes(x=treat, y=log(count), colour= HardCoral)) + 
    geom_boxplot() + 
    facet_wrap( ~ cluster_id ,scales = "free_y")+
  ggtitle("Differentially expressed genes in functional category (RESISTANCE) in treatment versus control")

plot_portiesR

# For genes classify as CELLULAR STRESS related

topCellularstressPorites<- DEG_174 %>% arrange(padj) %>% filter(CELLULAR.STRESS==TRUE) %>% select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) 
plot_portiesCS<- topCellularstressPorites %>%  
    pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
    left_join(sample_data) %>% 
    ggplot(aes(x=treat, y=log(count), colour= HardCoral)) + 
    geom_boxplot() + 
    facet_wrap( ~ cluster_id ,scales = "free_y") +
  ggtitle("Differentially expressed genes in functional category (CELLULAR STRESS) in treatment versus control")
plot_portiesCS
```


Manual Categories plots
```{r}
DE_174_database<- read.csv('~/Dropbox/natalia_competition/Porites_competition/gene_function/data/DE_174_database_underEdition2020.csv')

# We need to have the element only with the categories that we want
ThreeCats<- DE_174_database %>% 
  select("cluster_id", "Sensory.Nervous","Immunity","Cellular.Stress","log2FoldChange") %>% 
  mutate(updown = ifelse(log2FoldChange>=0,"up","down"))
```


```{r}
ComplexUpset::upset(ThreeCats,
      intersect = c("Sensory.Nervous","Immunity","Cellular.Stress"),
      base_annotations=list(
        'Intersection size'=intersection_size(
            counts=FALSE,
            aes=aes(fill=updown)
        )
      )
  )
```

