---
title: "Gene function analysis"
author: "Natalia Andrade and Ira Cooke"
date: "07/08/2017"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=FALSE, warning=FALSE, message=FALSE)
knitr::opts_chunk$set(cache=TRUE)
options(width = 60)

library(knitr)
library(DESeq2)
library(tidyverse)
library(readxl)
library(ggrepel)
library(ggplot2)
library(cowplot)
library(ggpubr)
# if(!require(devtools)) install.packages("devtools")
# devtools::install_github("krassowski/complex-upset")

#library(ComplexUpset)
```

Gene function analysis is based on the following datasets;
  - Functional annotations created in `01_annotate` for all clusters
  - Differential expression analysis (from `02_deseq.Rmd`) to select genes DE between control and treatment
  - Manual annotations created by curating automatic annotations along with literature searches for DE genes
  - K-means clustering groups which identify genes (Corset clusters) identified in the heatmap (see `04_polyp_activity.Rmd`)
  
```{r, include=FALSE}
# For the purpose of this analysis we first join 3 datasets together. 
# Dataset 1 is the differential expression statistics and normalized counts for all transcripts
# Dataset 2 is the annotation information for P.cylindrica transcripts
# Dataset 3 is a table of manual functional annotations for key genes

#
# Dataset 1
#
if (file.exists("cache/dds2.rds")) {
    dds2 <- read_rds("cache/dds2.rds")
} else {
  stop("Nothing in cache/dds2.rds. You must run 02_DESeq.Rmd first")
}

res_ct <- read_rds("cache/res_ct.rds")
vsd2 <- read_rds("cache/vsd2.rds")
quantifiable_clusters <- read_rds("cache/quantifiable_clusters.rds")

# (New) Normalized counts per sample (Dataset 1)
ct_res_ntd2 <- counts(dds2, normalized=TRUE)[res_ct$row,] %>%
  as.data.frame() %>%
  rownames_to_column("cluster_id") %>% 
  right_join(res_ct, by=c("cluster_id"="row"))
```

```{r}
#
# Dataset 2
#
# Transcriptome annotation 
# We select key columns here for readability
automatic_annotations <- read_rds("raw_data/annotated_clusters_porites.rda") %>% 
  dplyr::select(cluster_id,transcript_id,RNAMMER,Pfam,SignalP,TmHMM,eggnog,Kegg,ipro_pfam,ipro_pfam_desc,ipr_go,best_swissprot,Entry,Protein_names,Gene_names,Organism,Keywords,swissprot_go,plutea_id)

#colnames_to_exclude <- colnames(automatic_annotations)[-1]

# Dataset 3
#
# Manual annotation
manual_annotations <- read_excel("raw_data/manual_annotations_ira.xlsx",na = "NA") %>% 
  dplyr::select(-log2FoldChange)
```


```{r join_datasets}
# Joining all three datasets for the DEG
#
annotated_DEG <- ct_res_ntd2 %>% 
  left_join(automatic_annotations) %>% 
  left_join(manual_annotations)

annotated_quantifiable <- quantifiable_clusters %>% 
    left_join(automatic_annotations) 
```

Our focus initially is on the genes differentially expressed between control and treatment. Raw (normalised) counts for a handful of the top genes are plotted here as a sanity check to ensure that they look genuinely differentially expressed.

```{r}
sample_data <- read_csv("raw_data/Samples_data.csv",trim_ws = TRUE)
colnames(sample_data) <- c("ID","Tank","HardCoral",'treat',"SoftCoralControl","PdvsL_Other")


# For genes DE ONLY in  contrast btw Control and treatment

top_genes <- annotated_DEG %>% slice_min(padj, n=9, with_ties=F) %>%  pull(cluster_id)

plot_genes <- function(gene_list){
  annotated_DEG %>%
    filter(cluster_id %in% gene_list) %>% 
    select(cluster_id, PdLa,PdLb, PdLc,PdLe, PdC, PfLa, PfLb, PfLc, PfLd, PfLe, PfC) %>% 
    pivot_longer(c(-cluster_id),names_to = "ID",values_to="count") %>%
    left_join(sample_data) %>% 
    ggplot(aes(x=treat, y=log(count), colour=HardCoral)) +
    geom_jitter(aes(shape=SoftCoralControl), height=0, width=0.1) +
    facet_wrap(~cluster_id,ncol = 3,scales = "free_y") 
}


plot_genes(top_genes)
```



```{r}
#row_anno_data <- read_rds("cache/row_anno_data.rds")
row_anno_data <- readxl::read_excel("raw_data/annotated_DEG_hm_ira.xlsx",na = "NA") %>% 
  mutate(Secreted = ifelse(!is.na(secreted),"SE","FALSE")) %>% 
  mutate(Immunity = ifelse(!is.na(`immune response`),"I","FALSE")) %>% 
  mutate(Immunity_CellCell_Apoptosis = ifelse(!is.na(`immune response`) | !is.na(apoptosis) | !is.na(`cell-cell interactions`),"ICA","FALSE")) %>% 
  mutate(Stress = ifelse(!is.na(`ROS stress`) | !is.na(`unfolded protein response`),"SR","FALSE")) %>% 
  mutate(StressType = ifelse(!is.na(`ROS stress`),"ROS",NA)) %>% 
  mutate(StressType = ifelse(!is.na(`unfolded protein response`),"UPR",StressType)) %>%   
  mutate(CellCell = ifelse(!is.na(`cell-cell interactions`),"CC","FALSE")) %>%   
  mutate(Apoptosis = ifelse(!is.na(apoptosis),"A","FALSE")) %>% 
  mutate(Toxin = ifelse(!is.na(toxin),"T","FALSE")) %>% 
  mutate(GPCR = ifelse(!is.na(GPCR),"G","FALSE")) %>% 
      dplyr::select(cluster_id,Secreted,Immunity=Immunity,Toxin,Stress,Apoptosis,log2FoldChange,evidence_level,StressType,display_name) %>% as.data.frame()

fanno_data <- row_anno_data %>% 
  arrange(log2FoldChange) %>% 
  mutate(rank=-(row_number()-40)) %>% 
  pivot_longer(!any_of(c("cluster_id","log2FoldChange","rank","evidence_level","Toxin","StressType","display_name"))) %>% 
  filter(value!=FALSE) %>% 
  filter(name %in% c("Secreted","Immunity", "Stress"))
fanno_data[(fanno_data$name=="Secreted") & (fanno_data$evidence_level>1),'display_name'] <- ""
fanno_data$name <- factor(fanno_data$name,levels=c("Secreted","Immunity","Stress"))

fanno_data_toxins <- fanno_data %>% filter(Toxin=="T")
```

Next we plot the relative expression for genes in key categories, Immunity, Stress and Secreted (Toxins).  We use ranks in this plot because some of these genes have extreme outlying fold change values (due to being abundant in a few samples and absent in others) and would distort a more conventional plot such as a volcanoplot.  

```{r}

label_data <- fanno_data %>% select(display_name,rank) %>% distinct() %>% filter(display_name!="") %>% filter(!is.na(display_name))

pr <- fanno_data %>% 
  group_by(rank) %>% 
  mutate(prop=1/n()) %>% 
  ungroup() %>% 
  ggplot(aes(x=rank)) + 
  geom_col(aes(fill=name,y=prop),color="black") + theme_pubclean() + theme(legend.title = element_blank(), legend.position = "bottom") + xlab("Expression Rank") + 
  ylab("") + 
  geom_vline(aes(xintercept=0)) +

  #  geom_point(aes(y=0.5),size=1,color="black") + 
  geom_point(aes(y=1.1),data=fanno_data_toxins,size=2,color="blue") +
  geom_label(data=label_data,aes(label=display_name),y=1.5,size=3) +
  expand_limits(y=1.6) +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.line.x = element_line(), panel.grid.major.y = element_blank())

pm <- fanno_data %>% 
  ggplot(aes(x=rank)) + 
  geom_density(aes(color=name)) + theme_pubclean() + 
  guides(color=FALSE) + xlab("") + 
  theme(axis.line.x = element_blank(), axis.text.x = element_blank(),axis.ticks.x = element_blank())

library(cowplot)
plot_grid(pm,pr,nrow = 2, align = "v")

```

Figure 3: Ranking distribution of differentially expressed transcripts under non-contact competition by functional category: “Stress” in blue, “Immunity” in green and “Secreted” in red. Individual transcripts are shown as vertical bars positioned along the horizontal axis (points) according to the rank of their effect size (log2 fold change under competition compared with controls). Positive ranks represent transcripts up-regulated in competition compared to controls.  Transcripts belonging to multiple categories are shown as multi-coloured bars. The density of genes in each category are shown as smoothed lines above the bar plot. Blue points represent putative toxins. Category “Stress” includes transcripts associated with response to reactive oxygen species as well as those involved in the unfolded protein response. Labelled points correspond to transcripts with strong evidence of homology to named proteins (CAT: calase, AOSL: allene oxide synthase-lipoxygenase, SYT: synaptotagmin-like, TNFR: tumor necrosis factor receptor, CALR: calreticulin).

```{r}
ggsave("figures/functions2.png",width = 85,height = 120, units = "mm", dpi = 300)
ggsave("figures/functions2.pdf",width = 85,height = 120, units = "mm", dpi = 300)
```


