---
title: "Kraken analysis"
output: github_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
library(tidyverse)
library(cowplot)
```
To identify potential sources of contamination we profiles the taxonomic composition of raw RNA-seq reads using Kraken. Although Kraken can perform very rapid classification of reads it relies on a pre-built database based on available genomic sequences.  These databases are available for download for bacterial and viral sequences but need to be built separately for Eukaryotes.  Unfortunately it was not possible to include a large number of Eukaryotic genomes in this process due to the huge computational resources required.  We therefore built a custom database for approximate classification of reads including the following;

- Kraken standard bacteria
- *Porites lutea* (Coral) genome
- *Drosophila melanogaster* (Fruit fly) genome
- *Lottia gigantea* (Gastropod) genome
- *Amphimedon queenslandica* (Sponge) genome
- *Acropora digitifera* (Coral) genome
- *Amphibalanus amphitrite* (Barnacle) genome
- *Cladocopium* (Clade C - Algal Symbiont) genome
- *Breviolum* (Clade B - Algal Symbiont) genome
- *Durusdinium* (Clade D - Algal Symbiont) genome
- *Fugacium* (Clade F - Algal Symbiont) genome
- *Symbiodinium* (Clade A - Algal Symbiont) genome

Kraken 1 was used to build a database with kmer size 31 using these genomic resources.  Then raw RNA sequencing reads from each sample were classified and used to generate summaries of the LCA for each read in mpa format. 



```{r}
sample_data_raw <- read_csv("raw_data/Samples_data.csv",trim_ws = TRUE) %>% 
  unite("sample",HardCoral,Tanks,remove = FALSE)

genome_mpa31_files <- list.files("hpc/kraken/kraken_mpa/",pattern = "*.mpa",full.names = TRUE)

read_mpa <- function(path){
  s <- basename(path) %>% str_extract("[^\\.]+")
  mpa_data <- read_tsv(path,col_names = c("taxonomy","read_count"),col_types = cols()) %>% 
    add_column(sample=s) 
  mpa_data
}

genome_mpa31_data <- do.call(rbind,lapply(genome_mpa31_files,read_mpa)) %>% add_column(kmer="g31")

mpa_data <- genome_mpa31_data %>% 
  left_join(sample_data_raw) %>% 
  mutate(lowest_taxon = str_extract(taxonomy,"[^\\|]*$")) %>% 
  separate(lowest_taxon,c("level","taxon"),sep = '__', remove=FALSE)
```

The figure below summarises the LCA of reads. Key samples are shown with dots to the left of bars, with black dots indicating samples with known barnacle contamination and red dots show samples in group 4 (See main manuscript Figure 2). 

Since Coral and Symbiodiniaceae reads are so dominant the lower level contaminants are not visible in the top plot.  Three samples in the top plot (PfC, PdLd and PfLa) show some contamination with bacteria but this is still a relatively minor component of the overall read volume.  

The lower plot shows reads after excluding these dominant taxa.  This clearly reveals the three Barnacle containing samples as having a large number of reads with LCA as Hexanauplia (Barnacles).  The remaining samples all showed a small number of Gastropod and Sponge assigned reads as well as several thousand classified under Insecta.  Since the numbers of reads involved here are quite low this probably reflects some misclassification due to the highly incomplete nature of our kraken database.  Importantly, the key samples in group 4 show LCA profiles that were very similar to the other samples.  This suggests that DEG patterns in those samples are unlikely to have arisen due to a contaminant in those samples.  

```{r}
plot_data <- mpa_data %>% 
  filter(!grepl(taxonomy,pattern="d__Bacteria")) %>% 
  filter(level=="c") %>% 
  filter(read_count>100) %>% 
  rbind((mpa_data %>% 
  filter(grepl(taxonomy,pattern="^d__Bacteria$"))))

barnacle_samples <- c("PfLc","PdLe","PdLa")
group4_samples <- c("PfLd","PdLb","PfLa")

pA <- plot_data %>% 
  filter(read_count>100000) %>% 

    ggplot(aes(x=ID)) + geom_col(aes(y=read_count/1e6,fill=taxon)) + coord_flip() + xlab("") +
    geom_point(data = data.frame(ID=barnacle_samples),aes(y=-1)) +
  geom_point(data = data.frame(ID=group4_samples),color="red",aes(y=-0.5)) +
  ylab("Read Count  (Millions)") + theme_cowplot() + theme(legend.title = element_blank())

pC <- plot_data %>% 
  filter(read_count<1e6) %>% 
  filter(!grepl(taxonomy,pattern="d__Bacteria")) %>% 
  ggplot(aes(x=ID)) + 
  geom_col(aes(y=read_count/1000,fill=taxon)) + 
  geom_point(data = data.frame(ID=barnacle_samples),aes(y=-2)) +
  geom_point(data = data.frame(ID=group4_samples),color="red",aes(y=-1)) +
    coord_flip() + xlab("") + ylab("Read Count  (Thousands)") + theme_cowplot() + theme(legend.title = element_blank())


plot_grid(pA,pC,ncol=1)
```


```{r}
clade_names <- c("A"="Symbiodinium","B"="Breviolum","C"="Cladocopium","D"="Durusdinium","F"="Fugacium","Host"="Porites", "Barnacle"="Barnacle")

symbiodinium_data <- mpa_data %>% 
  filter(grepl(pattern = "Symbiodinium_sp_clade",taxonomy) | lowest_taxon == "o__Scleractinia") %>% 
  mutate(clade = str_match(taxonomy,pattern = "clade_([ABCDF])")[,2]) %>% 
  mutate(clade = ifelse(grepl( pattern = "o__Scleractinia", taxonomy),"Host",clade)) %>% 
    mutate(clade_name = clade_names[clade]) %>% 
 mutate(clade_name = ifelse(is.na(clade_name),"Porites",clade_name)) %>% 
  left_join(sample_data_raw)
```

Finally we examined the breakdown of Symbiont reads into components for the different Symbiodiniaceae genera.  This revealed that the dominant algal symbiont in this case was Cladocopium. 

```{r}
library(ggrepel)
library(ggpubr)

ggplot(symbiodinium_data ,aes(x=ID,y=read_count/1e6)) + 
  geom_col(aes(fill=reorder(clade_name, read_count))) +  theme_pubclean() +
  xlab("") + ylab("Read Count (Millions)") + 
  theme(axis.text.x = element_text(angle=90), axis.ticks.x = element_blank(), legend.title = element_blank(), legend.position = "right")
```
