---
title: "Genotyping"
output: github_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
```

To check that the two colonies of Porites used this study were indeed distinct genotypes we performed genotyping and relatedness analysis on all samples. 

First we removed clusters with multiple transcripts because these are indicative of potential alternative splicing or multiple gene copies and are more likely to produce incorrect read mappings and genotypes. 

In addition we removed clusters that were not expressed with at least a count of 3 in all 12 samples.  

After applying these filtering steps a list of 17093 good transcripts for genotyping were obtained. 

```{r, include=FALSE}
counts <- read_tsv("hpc/corset/03-counts.txt") %>% 
  column_to_rownames(var="X1")

good_counts <- counts[(rowSums(counts>3)==12),]

clusters <- read_tsv("hpc/corset/03-clusters.txt",col_names = c("transcript_id","cluster_id")) %>% 
  group_by(cluster_id) %>% 
  mutate(n_transcripts=n()) %>% 
  filter(n_transcripts==1) %>% 
  filter(cluster_id %in% rownames(good_counts))

write_tsv(clusters %>% ungroup %>% select(transcript_id),file="hpc/genotypes/good_clusters.txt", )
```
This was then used to create a bed file and read alignments overlapping this bed file were extracted for input to angsd. We then ran angsd to call SNPs and calculte genotype likelihoods.Analysis with angsd resulted in a total of 30289 SNPs.  These were then used for analysis with ngsRelate. Relevant commands are in;

  - [01_prepare_inputs.sh](hpc/genotyping/01_prepare_inputs.sh)
  - [02_angsd.sh](hpc/genotyping/02_angsd.sh)
  - [03_ngsrelate.sh](hpc/genotyping/03_ngsrelate.sh)


Although ngsRelate calculates many statistics the key statistic for our purposes is the relatedness statistic, theta.  For a diploid organism this statistic has a value of 0.5 for clones and should approach 0 for unrelated individuals.  In the figure below we show this value for all pairs of samples where samples numbered 0-5 are Pd and those numbered 6-11 are Pf.  It confirms our assumption that fragments from the same colony are clones (light blue) and fragments from different colonies have distinct (unrelated) genotypes.

```{r, message=FALSE}
res <- read_table2("hpc/genotypes/porites_relate")

ggplot(res) + geom_tile(aes(x=a,y=b,fill=theta))

# 0.5 - Monozygotic twins
# 0 - unrelated
```

